services/sso/src/main/java/software.amazon.awssdk.services.sso/internal/SsoCredentialsUtils.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sso.internal;
+
+import static software.amazon.awssdk.profiles.ProfileProperty.SSO_ACCOUNT_ID;
+import static software.amazon.awssdk.profiles.ProfileProperty.SSO_REGION;
+import static software.amazon.awssdk.profiles.ProfileProperty.SSO_ROLE_NAME;
+import static software.amazon.awssdk.profiles.ProfileProperty.SSO_START_URL;
+import static software.amazon.awssdk.utils.JavaSystemSetting.USER_NAME;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.util.Map;
+import javax.xml.bind.DatatypeConverter;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.profiles.Profile;
+import software.amazon.awssdk.services.sso.SsoClient;
+import software.amazon.awssdk.services.sso.auth.AwsSsoCredentials;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsRequest;
+import software.amazon.awssdk.services.sso.model.GetRoleCredentialsResponse;
+import software.amazon.awssdk.services.sso.model.RoleCredentials;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.internal.SystemSettingUtils;
+
+@SdkProtectedApi
+public final class SsoCredentialsUtils {
+
+    private static final String TOKEN_DIRECTORY = "/Users/" + SystemSettingUtils.resolveSetting(USER_NAME).get() +
+                                                  "/.aws/sso/cache/";
+
+    private Profile profile;
+    private SsoClient ssoClient;
+
+    private String ssoRegion;
+    private String ssoAccountId;
+    private String ssoRoleName;
+    private String ssoStartUrl;
```

Comment:
These not being final makes me nervous about thread-safety. The first two can easily be made final, but the next 4 only happen when we do getUpdatedCredentials. Is there any way to just do the work of getUpdatedCredentials at initialization and create a new `SsoCredentialsUtils` (there's probably a better name) whenever we need a refresh?